src/util/HashSet.h

```diff
@@ -32,4 +54,142 @@ template <class T,
+        memoryUsed_{MemorySize::bytes(0)},
+        sizeGetter_{sizeGetter},
+        currentSlotSize_{0} {
+    // Once the hash set is initialized, calculate the initial memory
+    // used by the slots of the hash set
+    updateSlotArrayMemoryUsage();
+  }
+
+  ~CustomHashSetWithMemoryLimit() { decreaseMemoryUsed(memoryUsed_); }
+
+  // Try to allocate the amount of memory requested
+  void increaseMemoryUsed(ad_utility::MemorySize amount) {
+    memoryLeft_.ptr()->wlock()->decrease_if_enough_left_or_throw(amount);
+    memoryUsed_ += amount;
+  }
+
+  // Decrease the amount of memory used
+  void decreaseMemoryUsed(ad_utility::MemorySize amount) {
+    memoryLeft_.ptr()->wlock()->increase(amount);
+    memoryUsed_ -= amount;
+  }
+
+  // Update the memory usage for the slot array if the bucket count changes.
+  // This function should be called after any operation that could cause
+  // rehashing. When the slot count increases, it reserves additional memory,
+  // and if the slot count decreases, it releases the unused memory back to the
+  // memory tracker.
+  void updateSlotArrayMemoryUsage() {
+    size_t newSlotSize = hashSet_.bucket_count();
+    if (newSlotSize != currentSlotSize_) {
+      if (newSlotSize > currentSlotSize_) {
+        ad_utility::MemorySize sizeIncrease =
+            ad_utility::MemorySize::bytes(slotMemoryCost) *
+            (newSlotSize - currentSlotSize_);
+        increaseMemoryUsed(sizeIncrease);
+      } else {
+        ad_utility::MemorySize sizeDecrease =
+            ad_utility::MemorySize::bytes(slotMemoryCost) *
+            (currentSlotSize_ - newSlotSize);
+
+        decreaseMemoryUsed(sizeDecrease);
+      }
+    }
+    currentSlotSize_ = newSlotSize;
+  }
+
+  // Insert an element into the hash set. If the memory limit is exceeded, the
+  // insert operation fails with a runtime error.
+  std::pair<typename HashSet::iterator, bool> insert(const T& value) {
+    MemorySize size =
+        sizeGetter_(value) + ad_utility::MemorySize::bytes(sizeof(T));
+    increaseMemoryUsed(size);
+
+    const auto& [it, wasInserted] = hashSet_.insert(value);
+
+    if (!wasInserted) {
+      decreaseMemoryUsed(size);
+    }
+
+    updateSlotArrayMemoryUsage();
+    return std::pair{it, wasInserted};
+  }
+
+  // _____________________________________________________________________________
+  void erase(const T& value) {
+    auto it = hashSet_.find(value);
+    if (it != hashSet_.end()) {
+      MemorySize size =
+          sizeGetter_(*it) + ad_utility::MemorySize::bytes(sizeof(T));
+      hashSet_.erase(it);
+      decreaseMemoryUsed(size);
+      updateSlotArrayMemoryUsage();
+    }
+  }
+
+  // _____________________________________________________________________________
+  void clear() {
+    hashSet_.clear();
+    // Release all node memory
+    decreaseMemoryUsed(memoryUsed_);
+
+    // Update slot memory usage based on the new bucket count after clearing
+    size_t newSlotSize = hashSet_.bucket_count();
```

Comment:
As stated, it should be possible (with a corresponding allocator which can be written), 
to get the exact size of everything but the external storage of the `string`s after you have performed an operation.